### PR TITLE
Update libsspwin32

### PR DESCRIPTION
**Pull request by yadij**
Move library files to lib/sspi/ for better modularity.

Add SQUID_CHECK_WIN32_SSPI autoconf test for dependency
checking by helpers.

Add missing HAVE_FOO_H wrappers around includes.
Fixes Squid coding style compliance.

Remove unnecessary __cplusplus protections.

Remove many redundant includes from helpers code.

Remove CygWin hack which is unreachable with precompiler
conditions.

Fix several structure initialization compiler errors.

### CODE REVIEW

**Comment by yadij** [comment]
This is draft because for me the library does not yet build. PR #1085 changed the C-style casting `(X) GetProcAddress(...)` to `reinterpret_cast<X>(GetProcAddress(...))` but neither form builds for me in GCC-12 cross-compile. That should be resolved before this merges, but not required - we have a lot of good progress here either way.

**Comment by yadij** [comment]
Also, this requires PR #1341 to be merged first (changes are included here).

**Review by rousskov — CHANGES_REQUESTED** [review]
> Also, this requires PR https://github.com/squid-cache/squid/pull/1341 to be merged first 

Agreed. Please ping me when that is settled, and I will review.

**Comment by yadij** [comment]
Since PR #1341 is seemingly stalled on idealistic pedantry I have rewritten this PR to not depend on it.

**Comment by kinkie** [comment]
Would you mind including some output from a configure run on mingw? The code looks good enough to me but it's the easiest way to double check it :)

File: acinclude/win32-sspi.m4
Comment (rousskov) [inline acinclude/win32-sspi.m4:14]: PR build [log](https://github.com/squid-cache/squid/suites/13813788092/artifacts/766560267):

```
checking whether Windows SSPI is available... checking for w32api/windows.h... no
checking for windows.h... (cached) no

configure: helper auth/basic/SSPI ... found but cannot be built
```

I see a few problems in the above output (but please ignore any specifics that are inaccurate -- no need to argue about these low-level details):

1. The first line appears to contain what should be two separate lines.
2. The outcome for "checking whether Windows SSPI is available..." (or "checking for w32api/windows.h...") check is not visible.
3. There is an empty line after "checking for windows.h... (cached) no"

Official build [log](https://github.com/squid-cache/squid/suites/13796614278/artifacts/765425688) looks much better:

```
checking for w32api/windows.h... no
checking for windows.h... (cached) no
configure: helper auth/basic/SSPI ... found but cannot be built
```

Please fix and check for other problems by comparing "before" and "after" ./configure output.

File: lib/sspi/sspwin32.cc
Comment (rousskov) [inline lib/sspi/sspwin32.cc:19]: To improve (IMO) readability, I recommend using the following (more prevalent in Squid code) formatting for this line:

```suggestion
template <typename T>
```

File: lib/sspi/sspwin32.cc
Comment (rousskov) [inline lib/sspi/sspwin32.cc:21]: In C++, `foo_cast<>` means "cast to foo". This function does not cast to FARPROC. It casts _from_ FARPROC.

I would also use CamelCase for this helper name so that the readers of the code using it know that it is not a standard cast:

```suggestion
FarProcToFunction(FARPROC in)
```

or, to follow Windows naming scheme,

```suggestion
FarProcToProc(FARPROC in)
```

or, simply,

```suggestion
ToFunction(FARPROC in)
```

or

```suggestion
ToProc(FARPROC in)
```

As usual, I do not insist on any specific spelling.

File: lib/sspi/sspwin32.cc
Comment (rousskov) [inline lib/sspi/sspwin32.cc:23]:
```
warning C4068: unknown pragma 'GCC'
```

To protect us from compiler warnings similar to the above, please wrap these pragmas in #ifdefs. There are examples in the official code.

AFAICT, we can assume that modern clang and GCC versions support these pragmas but compilers like MSVC [do not](https://godbolt.org/z/W51oT6rKn). Adding a `__GNUC__` wrapper helps in my tests, but there may be a better way.

**Review by rousskov — CHANGES_REQUESTED** [review]

File: lib/sspi/sspwin32.cc
Comment (yadij) [reply lib/sspi/sspwin32.cc:21]: > In C++, `foo_cast<>` means "cast to foo". 

WTF? that is a strange interpretation. It does not make any sense when applied to the standard X_cast operators.

It is supposed to be  ```[mechanism]_cast<foo>(expression)``` in this case the mechanism is Windows FARPROC type alias (inheritance, interfacing, puny-type, functor, function pointer, ... whatever it is).

> This function does not cast to FARPROC. It casts _from_ FARPROC.

Indeed, it casts toward the **type** template parameter.

File: lib/sspi/sspwin32.cc
Comment (yadij) [reply lib/sspi/sspwin32.cc:23]: Where did you get that warning from?

The ```GCC``` prefix is supposed to be standard way of defining the rest of the line as GCC-only and to be ignored by all other C++ compilers.

File: acinclude/win32-sspi.m4
Comment (yadij) [reply acinclude/win32-sspi.m4]: Fixed in [de3c084](https://github.com/squid-cache/squid/pull/1348/commits/de3c0845687cb525af643f1b55fec2870180ed3a)

Output now looks like:
```
checking for w32api/windows.h... no
checking for windows.h... (cached) yes
checking for wchar.h... (cached) yes
checking for tchar.h... yes
checking for ntsecapi.hhh... no
checking for security.h... yes
checking for sspi.h... yes
configure: helper auth/basic/SSPI ... found but cannot be built
```

File: lib/sspi/sspwin32.cc
Comment (rousskov) [reply lib/sspi/sspwin32.cc:21]: > Alex: In C++, foo_cast<> means "cast to foo". This function does not cast to FARPROC. It casts from FARPROC.

> Amos: It is supposed to be `[mechanism]_cast<foo>(expression)`

Agreed. I was wrong -- I went too far in that generalization. Sorry.

I still think we should use CamelCase for this helper name so that the readers of the code using it know that it is not a standard cast, but I do not insist on this change.

File: lib/sspi/sspwin32.cc
Comment (rousskov) [reply lib/sspi/sspwin32.cc:23]:
```
warning C4068: unknown pragma 'GCC'
```
 
> > Alex: AFAICT, we can assume that modern clang and GCC versions support these pragmas but compilers like MSVC [do not](https://godbolt.org/z/W51oT6rKn).

> Amos: Where did you get that warning from?

The problem is general, but that particular warning came from an MSVC test that was linked from the original change request: https://godbolt.org/z/W51oT6rKn

> The GCC prefix is supposed to be standard way of defining the rest of the line as GCC-only and to be ignored by all other C++ compilers.

It _is_ ignored, with a warning.

As for GCC prefix being "standard way", I see no such language/mechanism in pragma [documentation](https://en.cppreference.com/w/cpp/preprocessor/impl). There is a requirement to _ignore_ pragmas that are not recognized. AFAICT, that requirement does not prohibit issuing warnings (or supporting GCC pragmas in a non-GCC compiler for that matter).

BTW, these problems with pragmas is one of the many examples illustrating why I consider defaulting to `-Werror` a mistake, but this change request is not about that.

File: acinclude/win32-sspi.m4
Comment (rousskov) [inline acinclude/win32-sspi.m4:12]: The description got stale:

```suggestion
# Checks whether the Windows SSPI requirements are available and work.
# Sets squid_cv_win32_sspi to "yes" or "no".
# The argument is the commands to run on success.
```

The above suggestion also removes an empty trailing line in that comment. I do not insist on that removal because we are very inconsistent with regard to these comments styling (including `#`, `##`, and `dnl` variants and these empty lines), but I think it is better to converge towards no empty lines because our C++ descriptions do not use them and because they introduce an arguably unwanted distance from the thing being described, increasing the chances of description/code separation.

I also recommend rephrasing "requirements are available and work", but I do not know enough about SSPI to suggest specific improvements -- the macro code does not make the "work" part obvious (to me).

File: acinclude/win32-sspi.m4
Comment (rousskov) [inline acinclude/win32-sspi.m4:15]: If possible, please move this assignment inside the second AC_CHECK_HEADERS(). I assume the code "works" as is, but this assignment looks like the wrong "default" here IMHO.

**Review by rousskov — CHANGES_REQUESTED** [review]

File: lib/sspi/sspwin32.cc
Comment (yadij) [reply lib/sspi/sspwin32.cc:24]: Fixed in [9a918aa](https://github.com/squid-cache/squid/pull/1348/commits/9a918aaaacb1a9d5c0a7c834a591c7cfaf5a6005)

**Comment by yadij** [comment]
I think we are up-to-date now. Please re-review.

File: acinclude/win32-sspi.m4
Comment (rousskov) [reply acinclude/win32-sspi.m4:13]: Right, I only looked at the lines below the `yes` assignment, so my "second" AC_CHECK_HEADERS is your "third" -- we are talking about the same AC_CHECK_HEADERS() line that is described in PR code as "required headers".

> the if-found and if-not-found cases get run for each of them.

Ah, I did not realize that! I consider this change request addressed, thank you.

**Review by rousskov — COMMENTED** [review]
> PR description: This adds AC_CACHE to optimize build performance.

Please update the PR description to match the current code. I did not check the whole description, but the sentence quoted above is stale.

I see no problems in the current PR code, but I have not checked most of the low-level details.

**Comment by yadij** [comment]
Description refreshed.

@rousskov, are you wishing to do that low-level review? if not please mark as cleared.

**Comment by rousskov** [comment]
> @rousskov, are you wishing to do that low-level review? if not please mark as cleared.

Cleared for merge. FWIW, my "did not review X" disclaimers do not imply that I intend to review X (unless I explicitly say otherwise; in which case I would block the PR or self-request a review). That disclaimer is there primarily to let the PR author know that they are on their own with regard to X.

**Pull request closed without merging**
